File: app/src/main/java/ecarrara/eng/vilibra/domain/repository/BookCachedRepository.java
package ecarrara.eng.vilibra.domain.repository;

import android.text.TextUtils;

import ecarrara.eng.vilibra.domain.cache.Cache;
import ecarrara.eng.vilibra.domain.entity.Book;

/**
 * Concrete implementation of {@link BookRepository} that combines an implementation of
 * {@link Cache} with an implementation of {@link BookRepository}
 *
 */
public class BookCachedRepository implements BookRepository {

    private BookRepository bookRepository;
    private Cache<String, Book> bookCache;

    public BookCachedRepository(BookRepository bookRepository, Cache<String, Book> bookCache) {
        this.bookRepository = bookRepository;
        this.bookCache = bookCache;
    }

    @Override public Book byIsbn(String isbn) {
        Book book = this.bookCache.get(isbn);

        if(book.equals(Book.NO_BOOK)) {
            book = this.bookRepository.byIsbn(isbn);
            if(!book.equals(Book.NO_BOOK)) {
                String cacheKey = getCacheKeyForBook(book);
                if(TextUtils.isEmpty(cacheKey)) { // not possible to generate a cache key
                    book = Book.NO_BOOK;
                } else {
                    this.bookCache.put(cacheKey, book);
                }
            }
        }

        return book;
    }

    @Override public void add(Book book) {
        this.bookRepository.add(book);
    }

    private String getCacheKeyForBook(Book book) {
        String cacheKey = "";
        if(!TextUtils.isEmpty(book.getIsbn10())) {
            cacheKey = book.getIsbn10();
        } else if(!TextUtils.isEmpty(book.getIsbn13())) {
            cacheKey = book.getIsbn13();
        }
        return cacheKey;
    }
}

File: app/src/main/kotlin/br/eng/ecarrara/vilibra/book/data/BookCachedRepository.kt
package br.eng.ecarrara.vilibra.book.data

import br.eng.ecarrara.vilibra.book.data.datasource.BookLocalCache
import br.eng.ecarrara.vilibra.book.data.datasource.BookRemoteDataSource
import br.eng.ecarrara.vilibra.book.domain.BookRepository
import br.eng.ecarrara.vilibra.book.domain.entity.Book
import br.eng.ecarrara.vilibra.book.domain.entity.BookWithNotValidIsbnException
import br.eng.ecarrara.vilibra.core.data.RxCache
import io.reactivex.Single
import timber.log.Timber
import javax.inject.Inject

/**
 * Concrete implementation of [BookRepository] that combines an implementation of
 * [RxCache] with an implementation of [BookRepository]
 *
 */
class BookCachedRepository
@Inject constructor(
        private val bookRemoteDataSource: BookRemoteDataSource,
        private val bookLocalCache: BookLocalCache
) : BookRepository {

    override fun getByIsbn(isbn: String): Single<Book> {
        return Single
                .concat(bookLocalCache[isbn], getFromRemoteDataSourceAndStoreToCache(isbn))
                .doOnComplete { Timber.d("Current Thread: " + Thread.currentThread().name) }
                .firstOrError()
    }

    private fun getFromRemoteDataSourceAndStoreToCache(isbn: String) = bookRemoteDataSource
            .searchForBookBy(isbn)
            .flatMap { book -> add(book).andThen(bookLocalCache[book.isbn10]) }

    override fun add(book: Book) = bookLocalCache.put(getCacheKeyForBook(book), book)

    private fun getCacheKeyForBook(book: Book): String {
        with(book, {
            return when {
                isbn10.isNotEmpty() -> isbn10
                isbn13.isNotEmpty() -> isbn13
                else -> throw BookWithNotValidIsbnException()
            }
        })
    }
}
